### PR TITLE
Fix clippy lints

### PR DESCRIPTION
**Pull request by JoeyBF**
I went through the codebase and fixed all the lints. I don't know if some of these were optimizations but the few benches I ran showed no regressions

### CODE REVIEW

**Comment by adamreichold** [comment]
Rayon currently has an MSRV of 1.36.0 so that e.g. `mem::take` cannot be used in the code base as it was stabilised only in 1.40.0.

**Comment by JoeyBF** [comment]
Good point, I made sure it compiles with 1.36. I left some `#[allow(...)]` for when the MSRV gets raised in the future

**Review by cuviper — CHANGES_REQUESTED** [review]
I went ahead and fixed a few more in rayon-demo -- will push in a moment.

File: src/iter/mod.rs [hunk 1]
```diff
@@ -2877,6 +2877,23 @@ pub trait IndexedParallelIterator: ParallelIterator {
...
+    /// ```
+    /// use rayon::prelude::*;
+    ///
+    /// let par_iter1 = (0..100).into_par_iter();
+    /// assert!(!par_iter1.is_empty());
+    ///
+    /// let par_iter2 = (0..0).into_par_iter();
+    /// assert!(par_iter2.is_empty());
+    /// ```
+    fn is_empty(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (JoeyBF) [reply src/iter/mod.rs]: Good point, I added an `allow` and a comment about that

**Comment by cuviper** [comment]
Thanks!

bors r+

**Pull request merged**